warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldMissingFromSchemaVisitorTest.java

```diff
@@ -0,0 +1,328 @@
+package datawave.query.jexl.visitors;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.util.MockMetadataHelper;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ParseException;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Set;
+
+import static datawave.query.Constants.ANY_FIELD;
+import static datawave.query.Constants.NO_FIELD;
+import static org.junit.Assert.assertEquals;
+
+public class FieldMissingFromSchemaVisitorTest {
+    
+    // Special fields required by visitor.
+    private Set<String> specialFields = Sets.newHashSet(ANY_FIELD, NO_FIELD);
+    
+    /**
+     * Test query with two field:datatype pairs that both exist in the metadata helper.
+     */
+    @Test
+    public void testWithFieldsThatExist() throws ParseException {
+        String query = "FOO == 'bar' && FOO2 == 'bar'";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        
+        // Fields to datatypes for our MetadataHelper
+        Multimap<String,String> fieldsToDatatypes = HashMultimap.create();
+        fieldsToDatatypes.put("FOO", "datatype1");
+        fieldsToDatatypes.put("FOO2", "datatype2");
+        
+        // Setup MockMetadataHelper
+        MockMetadataHelper helper = new MockMetadataHelper();
+        helper.addFieldsToDatatypes(fieldsToDatatypes);
+        
+        // Setup datatype filter
+        Set<String> dataTypeFilter = Sets.newHashSet("datatype1", "datatype2");
+        
+        // Case 1 - check with a datatype filter
+        Set<String> actual = FieldMissingFromSchemaVisitor.getNonExistentFields(helper, script, dataTypeFilter, specialFields);
+        assertEquals(Collections.emptySet(), actual);
+        
+        // Case 2 - check with a null datatype filter, implies all fields.
+        actual = FieldMissingFromSchemaVisitor.getNonExistentFields(helper, script, null, specialFields);
+        assertEquals(Collections.emptySet(), actual);
+        
+        // Case 3 - check with an empty datatype filter, implies all fields.
+        actual = FieldMissingFromSchemaVisitor.getNonExistentFields(helper, script, Collections.emptySet(), specialFields);
+        assertEquals(Collections.emptySet(), actual);
+    }
+    
+    /**
+     * Test query with two field:datatype pairs, none of which does not exist in the metadata helper.
```

Comment:
None of which does _not_ exist?  So none of them do not exist in the metadata helper?  Or, both of which do not exist in the metadata helper?

Reply:
Should be "none of which exist". Fixed in next commit.